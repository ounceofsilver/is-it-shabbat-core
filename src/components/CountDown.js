import { Component } from 'react';
import PropTypes from 'prop-types';
import { DateTime } from 'luxon';

export default class CountDown extends Component {
	constructor(props) {
		super(props);
		this.state = {
			realNow: DateTime.local(),
		};
		this.startCountdown();
	}

	componentDidMount() {
		const { refreshInterval } = this.props;
		this.timerId = setInterval(
			() => this.tick(),
			refreshInterval,
		);
	}

	componentWillUnmount() {
		clearInterval(this.timerId);
	}

	getTime() {
		const { realNow } = this.state;
		const { start } = this.props;
		return start.plus(realNow - this.realStart);
	}

	startCountdown() {
		this.realStart = DateTime.local();
	}

	durationLeft() {
		const { end, start } = this.props;
		const { realNow } = this.state;
		return (end.diff(start)).minus(realNow - this.realStart);
	}

	update() {
		this.setState({
			realNow: DateTime.local(),
		});
	}

	tick() {
		const { callback } = this.props;
		this.update();
		if (this.durationLeft() <= 0) {
			callback(this.getTime());
			this.startCountdown();
			this.update();
		}
	}

	render() {
		const d = this.durationLeft();
		const { children } = this.props;
		return children(d);
	}
}
CountDown.propTypes = {
	start: PropTypes.instanceOf(DateTime).isRequired,
	end: PropTypes.instanceOf(DateTime).isRequired,
	refreshInterval: PropTypes.number,
	callback: PropTypes.func,
	children: PropTypes.func,
};
CountDown.defaultProps = {
	callback: () => {},
	children: () => {},
	refreshInterval: 100,
};
