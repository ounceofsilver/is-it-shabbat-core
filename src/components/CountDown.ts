import { DateTime } from 'luxon';
import React, { Component } from 'react';

declare const clearInterval;
declare const setInterval;

export class CountDown extends Component<{
	start: DateTime,
	end: DateTime,
	refreshInterval?: number,
	callback?: (now: DateTime) => void,
	children?: (now: DateTime) => JSX.Element,
}, {
	realNow: DateTime,
}> {
	private timerId?: number;
	private realStart?: DateTime;

	constructor(props) {
		super(props);
		this.state = {
			realNow: DateTime.local(),
		};
		this.startCountdown();
	}

	public componentDidMount = () => {
		this.timerId = setInterval(
			() => this.tick(),
			this.props.refreshInterval || 100,
		);
	}

	public render() {
		const d = this.durationLeft();
		const { children = () => null } = this.props;
		return children(d);
	}

	public componentWillUnmount() {
		clearInterval(this.timerId);
	}

	private getTime() {
		return this.props.start.plus(this.state.realNow - this.realStart);
	}

	private startCountdown() {
		this.realStart = DateTime.local();
	}

	private durationLeft() {
		const { end, start } = this.props;
		const { realNow } = this.state;
		return (end.diff(start)).minus(realNow - this.realStart);
	}

	private update() {
		this.setState({
			realNow: DateTime.local(),
		});
	}

	private tick() {
		const { callback } = this.props;
		this.update();
		if (this.durationLeft() <= 0) {
			if (this.props.callback) {
				this.props.callback(this.getTime());
			}
			this.startCountdown();
			this.update();
		}
	}
}
