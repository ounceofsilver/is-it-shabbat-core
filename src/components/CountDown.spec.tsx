import { Duration } from 'luxon';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { local } from '../test';
import { CountDown } from './CountDown';

describe('CountDown', () => {
	it('should render', () => {
		const childSpy = jest.fn().mockImplementation(
			(d: Duration) => <p>
				Hello
				{Math.floor(d.shiftTo('milliseconds').milliseconds / 1000)}
				s
			</p>,
		);
		const cbSpy = jest.fn();
		const props = {
			start: local(2018, 12, 24, 0, 1, 0),
			end: local(2018, 12, 24, 0, 1, 1),
			callback: cbSpy,
			refreshInterval: 500,
		};
		const wrapper = renderer.create(
			<CountDown {...props}>
				{childSpy}
			</CountDown>,
		);
		expect(wrapper.toJSON()).toMatchSnapshot();
	});

	it('should call child function with remaining duration', () => {
		jest.useFakeTimers();
		const childSpy = jest.fn().mockReturnValue(<p>Hello</p>);
		const cbSpy = jest.fn();
		const props = {
			start: local(2018, 12, 24, 0, 1, 0),
			end: local(2018, 12, 24, 0, 1, 1),
			callback: cbSpy,
			refreshInterval: 500,
		};
		renderer.create(
			<CountDown {...props}>
				{childSpy}
			</CountDown>,
		);

		// First cycle
		expect(childSpy).toHaveBeenCalled();
		childSpy.mockClear();

		// Next cycle
		jest.advanceTimersByTime(500);

		expect(childSpy).toHaveBeenCalled();
		expect(childSpy).toHaveBeenCalledWith(expect.any(Duration));
	});

	it('should work with default callback and no children', () => {
		jest.useFakeTimers();
		const props = {
			start: local(2018, 12, 24),
			end: local(2018, 12, 24, 0, 1),
			refreshInterval: 5000,
		};
		renderer.create(<CountDown {...props} />);
		jest.advanceTimersByTime(60 * 1000); // Hitting default callback
	});
});
