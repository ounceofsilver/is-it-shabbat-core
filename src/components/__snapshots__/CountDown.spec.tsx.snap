// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountDown should render 1`] = `
<p>
  Hello
  1
  s
</p>
`;
